### PR TITLE
fs/tmpfs: Handle the tail '/' correctly

### PR DESCRIPTION
**Pull request by xiaoxiang781216**
## Summary
and mkdir return EEXIST if the path already exist

## Impact
Handle the tail '/' robustly

### CODE REVIEW

**Review by Ouss4 — APPROVED** [review]

File: fs/tmpfs/fs_tmpfs.c [hunk 1]
```diff
@@ -455,6 +455,13 @@ static int tmpfs_find_dirent(FAR struct tmpfs_directory_s *tdo,
 {
   int i;
 
+  if (len && name[len - 1] == '/')
```
Comment (acassis) [inline fs/tmpfs/fs_tmpfs.c:458]: Why don't you do the same as you did for namelen:
if (len == 0)
  {
    return -EINVAL;
  }
else if (name[len - 1] == '/')

File: fs/tmpfs/fs_tmpfs.c [hunk 2]
```diff
@@ -992,14 +1026,21 @@ static int tmpfs_find_file(FAR struct tmpfs_s *fs,
                            FAR struct tmpfs_directory_s **parent)
 {
   FAR struct tmpfs_object_s *to;
+  size_t len;
   int ret;
 
+  len = strlen(relpath);
+  if (len && relpath[len - 1] == '/')
```
Comment (acassis) [inline fs/tmpfs/fs_tmpfs.c:1033]: Ditto

**Review by acassis — CHANGES_REQUESTED** [review]

File: fs/tmpfs/fs_tmpfs.c (hunk 1 above)
Comment (xiaoxiang781216) [reply fs/tmpfs/fs_tmpfs.c]: Done.

File: fs/tmpfs/fs_tmpfs.c (hunk 2 above)
Comment (xiaoxiang781216) [reply fs/tmpfs/fs_tmpfs.c:1040]: We need return fail here because it's invalid that file name contain the tail '/'.

Comment (acassis) [reply fs/tmpfs/fs_tmpfs.c:1040]: I mean: my point was the same:
if (len == 0)
 {
  return -EINVAL;
 }
else (relpath[len - 1] == '/') ...

Comment (xiaoxiang781216) [reply fs/tmpfs/fs_tmpfs.c]: Done.

**Review by acassis — APPROVED** [review]

**Comment by acassis** [comment]
@xiaoxiang781216 what you think of this modification:
```
+++ b/fs/tmpfs/fs_tmpfs.c
@@ -463,6 +463,8 @@ static int tmpfs_find_dirent(FAR struct tmpfs_directory_s *tdo,
     {
       /* Ignore the tail '/' */
 
+      name[len - 1] = '\0';
+
       if (--len == 0)
         {
           return -EINVAL;
@@ -473,8 +475,8 @@ static int tmpfs_find_dirent(FAR struct tmpfs_directory_s *tdo,
 
   for (i = 0;
        i < tdo->tdo_nentries &&
-       (strncmp(tdo->tdo_entry[i].tde_name, name, len) != 0 ||
-       tdo->tdo_entry[i].tde_name[len] != 0);
+       (tdo->tdo_entry[i].tde_name[len] != 0 ||
+        strncmp(tdo->tdo_entry[i].tde_name, name, len) != 0);
        i++);
 
   /* Return what we found, if anything */
```
It will save processing time, because most of time the strncmp will not be executed.

**Comment by xiaoxiang781216** [comment]
but, the length of tde_name may shorter than len and then the access may out of bound.

**Comment by acassis** [comment]
That is true, it only should be executed if s1 > s2. Unfortunately we don't have a tde_name_len :-)

**Pull request merged**